File: android/app/src/main/kotlin/org/gaindrive/android/data/SettingsStore.kt
package org.gaindrive.android.data

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import org.gaindrive.android.data.model.AlbumSort
import org.gaindrive.android.data.model.AudioQuality
import org.gaindrive.android.data.model.LibrarySection
import org.gaindrive.android.data.model.ThemeMode
import javax.inject.Inject
import javax.inject.Singleton

/**
 * App-wide preferences that are not tied to any one server.
 */
@Singleton
class SettingsStore @Inject constructor(
	private val dataStore: DataStore<Preferences>,
) {

	val themeMode: Flow<ThemeMode> = dataStore.data.map { prefs ->
		prefs[THEME]?.let { name -> runCatching { ThemeMode.valueOf(name) }.getOrNull() }
			?: ThemeMode.AUTO
	}

	suspend fun setThemeMode(mode: ThemeMode) {
		dataStore.edit { it[THEME] = mode.name }
	}

	/**
	 * What the library screens are showing: a server id, or
	 * [org.gaindrive.android.data.model.BrowseScope.ALL_STORED]. Null until the
	 * user has chosen; resolving it is [ServerSelection]'s job, since the stored
	 * choice may since have been disabled or removed.
	 */
	val browseScope: Flow<String?> = dataStore.data.map { it[SELECTED_SERVER] }

	suspend fun setBrowseScope(value: String) {
		dataStore.edit { it[SELECTED_SERVER] = value }
	}

	/**
	 * Which order the albums screen lists an artist's albums in.
	 *
	 * Kept per library section, because they are browsed for different
	 * reasons: a discography is chronological, while a film category is
	 * findable only by name. The caller names the section it drilled in from
	 * — this used to key on a stored "current mode" preference, which could
	 * disagree with the listing actually on screen.
	 */
	fun albumSort(section: LibrarySection): Flow<AlbumSort> =
		dataStore.data.map { AlbumSort.parse(it[albumSortKey(section)]) }

	suspend fun setAlbumSort(section: LibrarySection, sort: AlbumSort) {
		dataStore.edit { it[albumSortKey(section)] = sort.name }
	}

	/**
	 * Whether an album held on several servers collapses to one row.
	 *
	 * On by default. Anyone running two servers at once is likely to have the
	 * same album on both — a streaming collection beside locally downloaded
	 * copies of it — and seeing every one of them twice is the worse default.
	 * The badges on a collapsed row keep it honest about what was folded away.
	 */
	val mergeDuplicateAlbums: Flow<Boolean> =
		dataStore.data.map { it[MERGE_ALBUMS] ?: true }

	suspend fun setMergeDuplicateAlbums(enabled: Boolean) {
		dataStore.edit { it[MERGE_ALBUMS] = enabled }
	}

	/**
	 * How much audio the cache may hold. Pinned downloads can push it past this
	 * — see `PinAwareEvictor` — but nothing else may.
	 */
	val cacheMaxBytes: Flow<Long> =
		dataStore.data.map { it[CACHE_MAX_BYTES] ?: DEFAULT_CACHE_BYTES }

	suspend fun setCacheMaxBytes(bytes: Long) {
		dataStore.edit { it[CACHE_MAX_BYTES] = bytes }
	}

	/**
	 * Whether playing a track also stores it.
	 *
	 * On by default: replaying an album is the common case, and the cap plus
	 * automatic eviction means it cannot run away. Turning it off still leaves
	 * the cache readable, so pinned downloads keep working.
	 */
	val cacheOnPlay: Flow<Boolean> = dataStore.data.map { it[CACHE_ON_PLAY] ?: true }

	suspend fun setCacheOnPlay(enabled: Boolean) {
		dataStore.edit { it[CACHE_ON_PLAY] = enabled }
	}

	/**
	 * Behave as though there were no network, whatever the network says.
	 *
	 * A mode, not a display preference: it stops requests being made at all, so
	 * a slow or expensive connection costs nothing and no screen waits out a
	 * timeout before showing what is stored.
	 */
	val offlineMode: Flow<Boolean> = dataStore.data.map { it[OFFLINE_MODE] ?: false }

	suspend fun setOfflineMode(enabled: Boolean) {
		dataStore.edit { it[OFFLINE_MODE] = enabled }
	}

	/**
	 * Whether pinned downloads wait for an unmetered connection.
	 *
	 * On by default, and only applies to downloads: caching what you are
	 * already streaming costs no extra data, so gating that on Wi-Fi would
	 * penalise the mobile listener for nothing.
	 */
	val downloadUnmeteredOnly: Flow<Boolean> =
		dataStore.data.map { it[DOWNLOAD_UNMETERED_ONLY] ?: true }

	suspend fun setDownloadUnmeteredOnly(enabled: Boolean) {
		dataStore.edit { it[DOWNLOAD_UNMETERED_ONLY] = enabled }
	}

	/**
	 * What quality to fetch audio at — downloads, cache-on-play and plain
	 * streaming alike.
	 *
	 * One setting rather than one per connection type: everything played is
	 * cached, so a separate streaming quality would only mean storing something
	 * different from what a download of the same track produces, and then
	 * holding both.
	 *
	 * Stored as [AudioQuality.tag] so it is a single value — a format and a
	 * bitrate written separately could be observed half-applied.
	 */
	val audioQuality: Flow<AudioQuality> = dataStore.data.map { prefs ->
		prefs[AUDIO_QUALITY]?.let(AudioQuality::parse) ?: AudioQuality.DEFAULT
	}

	suspend fun setAudioQuality(quality: AudioQuality) {
		dataStore.edit { it[AUDIO_QUALITY] = quality.tag }
	}

	/**
	 * Whether a Chromecast fetching from the server itself is sent the file as
	 * it stands, rather than the streaming quality above.
	 *
	 * The setting exists because [audioQuality] answers a different question:
	 * what is worth sending to *this phone*, where a transcode saves battery,
	 * airtime and cache. A television pulling straight off the server over a
	 * wired LAN spends none of those, so the transcode there buys nothing and
	 * costs fidelity.
	 *
	 * It applies only to that route. A cast relayed through the phone crosses
	 * this phone's Wi-Fi and, when roaming, its VPN and mobile data — the
	 * situation the bridge exists for, and the last place to start sending
	 * FLAC. `CastUrls` is where that distinction is made.
	 *
	 * Off by default: turning it on changes what every existing cast sends, and
	 * on a weak access point that is a regression nobody asked for.
	 */
	val castOriginal: Flow<Boolean> = dataStore.data.map { it[CAST_ORIGINAL] ?: false }

	suspend fun setCastOriginal(enabled: Boolean) {
		dataStore.edit { it[CAST_ORIGINAL] = enabled }
	}

	/**
	 * Whether a video is played for its soundtrack alone.
	 *
	 * Backing out of the video surface does not do this — the whole picture
	 * still arrives over the network, and nothing is kept afterwards, because
	 * video deliberately never enters the byte cache. Asking the server for an
	 * audio `format` instead makes it send only the audio track (see
	 * `audio_only_request()` in `src/codecs.hh`), and what arrives is then an
	 * ordinary audio stream in every respect: cached on play, downloadable for
	 * offline, and seekable.
	 *
	 * Global rather than a per-track toggle, and that is what makes the caching
	 * work at all: a pin has to know, before anything is fetched, whether the
	 * bytes it stores for a film are its soundtrack or nothing. A toggle that
	 * moved per track would leave [org.gaindrive.android.data.cache.PinRepository]
	 * with no stable answer.
	 *
	 * Off by default: a video library is a video library until someone says
	 * otherwise.
	 */
	val videoAudioOnly: Flow<Boolean> = dataStore.data.map { it[VIDEO_AUDIO_ONLY] ?: false }

	suspend fun setVideoAudioOnly(enabled: Boolean) {
		dataStore.edit { it[VIDEO_AUDIO_ONLY] = enabled }
	}

	/**
	 * What the URL-fetch panel was last set to.
	 *
	 * All four are sticky between shares, which is what the web client does and
	 * for the same reason: several tracks going into one album is the ordinary
	 * case, and retyping the album for each of them is the whole friction the
	 * fields were added to remove. The panel's Clear button is the safeguard
	 * against a name outliving its usefulness — see `web/app.js`, whose comment
	 * notes the names "deliberately survive".
	 *
	 * The server is stored as a raw [org.gaindrive.android.data.model.ServerId]
	 * string, and may name one since removed or since stripped of its upload
	 * rights; resolving it against what is actually eligible is the panel's job.
	 */
	val fetchServer: Flow<String?> = dataStore.data.map { it[FETCH_SERVER] }

	suspend fun setFetchServer(value: String) {
		dataStore.edit { it[FETCH_SERVER] = value }
	}

	/** "audio" or "video". Audio by default: this is a music library. */
	val fetchAudio: Flow<Boolean> = dataStore.data.map { it[FETCH_AUDIO] ?: true }

	suspend fun setFetchAudio(audio: Boolean) {
		dataStore.edit { it[FETCH_AUDIO] = audio }
	}

	val fetchArtist: Flow<String> = dataStore.data.map { it[FETCH_ARTIST] ?: "" }

	val fetchAlbum: Flow<String> = dataStore.data.map { it[FETCH_ALBUM] ?: "" }

	/** Written together, because Clear has to be able to forget both at once. */
	suspend fun setFetchNames(artist: String, album: String) {
		dataStore.edit {
			it[FETCH_ARTIST] = artist
			it[FETCH_ALBUM] = album
		}
	}

	/**
	 * Whether the user has ever chosen a quality.
	 *
	 * Used once, at startup: someone upgrading with pins already downloaded
	 * should not silently have all of them re-fetched because the new default
	 * differs from what those bytes are. See `QualityMigration`.
	 */
	val audioQualityChosen: Flow<Boolean> =
		dataStore.data.map { it[AUDIO_QUALITY] != null }

	companion object {
		/** Big enough to be useful, small enough not to surprise anyone. */
		const val DEFAULT_CACHE_BYTES = 4L * 1024 * 1024 * 1024

		private val THEME = stringPreferencesKey("theme_mode")
		private val SELECTED_SERVER = stringPreferencesKey("selected_server")

		/** One key per library section, so the sections do not share an answer. */
		private fun albumSortKey(section: LibrarySection) =
			stringPreferencesKey("album_sort_${section.id}")

		private val MERGE_ALBUMS = booleanPreferencesKey("merge_duplicate_albums")
		private val CACHE_MAX_BYTES = longPreferencesKey("cache_max_bytes")
		private val CACHE_ON_PLAY = booleanPreferencesKey("cache_on_play")
		private val DOWNLOAD_UNMETERED_ONLY = booleanPreferencesKey("download_unmetered_only")
		private val OFFLINE_MODE = booleanPreferencesKey("offline_mode")
		private val AUDIO_QUALITY = stringPreferencesKey("audio_quality")
		private val CAST_ORIGINAL = booleanPreferencesKey("cast_original")
		private val VIDEO_AUDIO_ONLY = booleanPreferencesKey("video_audio_only")
		private val FETCH_SERVER = stringPreferencesKey("fetch_server")
		private val FETCH_AUDIO = booleanPreferencesKey("fetch_audio")
		private val FETCH_ARTIST = stringPreferencesKey("fetch_artist")
		private val FETCH_ALBUM = stringPreferencesKey("fetch_album")
	}
}

File: ios/GainDrive/Data/SettingsStore.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// Settings that are not about any one server.
///
/// Deliberately Foundation-only — the mapping from `ThemeMode` to SwiftUI's
/// `ColorScheme` lives in the UI layer, so the dependency direction stays
/// `UI → Data → Net`.
@MainActor
@Observable
final class SettingsStore {
	var themeMode: ThemeMode {
		didSet { defaults.set(themeMode.rawValue, forKey: Self.themeKey) }
	}

	/// The browse scope, as `BrowseScope.stored` — a server's UUID or the
	/// `"all"` sentinel. Kept as a string rather than as a `BrowseScope`
	/// because resolving it needs the current server list, which this type
	/// deliberately knows nothing about. `ServerSelection` does the resolving.
	var selectedServer: String? {
		didSet { defaults.set(selectedServer, forKey: Self.selectedServerKey) }
	}

	/// **On by default here, off by default on Android.** The two apps
	/// genuinely differ: `ios/PLAN.md` says "Albums merge, on by default" and
	/// `android/data/Merge.kt` says the opposite. Recorded so the next reader
	/// does not "fix" one of them into agreement.
	var mergeDuplicateAlbums: Bool {
		didSet { defaults.set(mergeDuplicateAlbums, forKey: Self.mergeAlbumsKey) }
	}

	/// One global quality, capped per track by *that track's* account ceiling —
	/// a queue spanning servers crosses caps at every boundary, so this is
	/// deliberately not per server.
	var audioQuality: AudioQuality {
		didSet { defaults.set(audioQuality.tag, forKey: Self.audioQualityKey) }
	}

	/// Stored music only — no server is contacted at all.
	///
	/// **A mode, not a display filter.** The repository skips the request
	/// rather than making it and hiding the answer: a manual offline mode on a
	/// flaky connection would otherwise be worse than useless, waiting out a
	/// full timeout before showing the mirror it could have shown at once.
	///
	/// It lives in the library picker rather than in Settings because it
	/// answers the same question the scope does, and because it is flipped
	/// before a flight rather than configured once.
	var offlineMode: Bool {
		didSet { defaults.set(offlineMode, forKey: Self.offlineKey) }
	}

	/// Which slice of the library was last showing, as `LibraryMode.id`.
	///
	/// Stored as the raw string rather than as a `LibraryMode`, so a value
	/// written by a build that knew a kind this one does not still parses —
	/// which is the same reason `LibraryMode` wraps a string at all. An
	/// unrecognised one falls back when the chip row is read.
	var libraryMode: String? {
		didSet { defaults.set(libraryMode, forKey: Self.libraryModeKey) }
	}

	/// Album order, **per slice**, so films can sit A–Z while a musician's
	/// albums stay chronological. A single global setting would make one of
	/// those two wrong every time the other was set.
	private(set) var albumSorts: [String: String] {
		didSet { defaults.set(albumSorts, forKey: Self.albumSortsKey) }
	}

	func albumSort(for mode: LibraryMode) -> AlbumSort {
		AlbumSort.parse(albumSorts[mode.id])
	}

	func setAlbumSort(_ sort: AlbumSort, for mode: LibraryMode) {
		albumSorts[mode.id] = sort.rawValue
	}

	/// How much downloaded audio may sit on the device.
	///
	/// **A ceiling on a refusal, not on an evictor** — for now. Nothing here is
	/// cached-on-play yet, so everything stored was explicitly asked for and
	/// nothing may be reclaimed; pinning past the cap is refused instead. When
	/// cache-on-play arrives this becomes the evictor's bound as well, and the
	/// refusal stays, because eviction cannot reclaim pinned bytes.
	///
	/// 4 GB to match Android's `DEFAULT_CACHE_BYTES`. Stored as a `Double`
	/// because `UserDefaults` has no `Int64` accessor and a music library
	/// passes what an `Int` is guaranteed to hold on a 32-bit device.
	var cacheCapBytes: Int64 {
		didSet { defaults.set(Double(cacheCapBytes), forKey: Self.cacheCapKey) }
	}

	static let defaultCacheCapBytes: Int64 = 4 * 1024 * 1024 * 1024
	static let cacheCapChoices: [Int64] = [
		1 * 1024 * 1024 * 1024,
		2 * 1024 * 1024 * 1024,
		4 * 1024 * 1024 * 1024,
		8 * 1024 * 1024 * 1024,
		16 * 1024 * 1024 * 1024,
	]

	@ObservationIgnored private let defaults: UserDefaults
	// Spelled as Android spells them in its DataStore, so the two apps
	// describe the same settings by the same names.
	private static let themeKey = "theme_mode"
	private static let selectedServerKey = "selected_server"
	private static let mergeAlbumsKey = "merge_duplicate_albums"
	private static let audioQualityKey = "audio_quality"
	private static let libraryModeKey = "library_mode"
	private static let albumSortsKey = "album_sort"
	private static let cacheCapKey = "cache_max_bytes"
	private static let offlineKey = "offline_mode"

	init(defaults: UserDefaults = .standard) {
		self.defaults = defaults
		let stored = defaults.string(forKey: Self.themeKey)
		themeMode = stored.flatMap(ThemeMode.init(rawValue:)) ?? .auto
		selectedServer = defaults.string(forKey: Self.selectedServerKey)
		// `bool(forKey:)` answers false for an absent key, so the default has
		// to be read through `object(forKey:)` or a fresh install would get
		// the opposite of what is intended.
		mergeDuplicateAlbums =
			defaults.object(forKey: Self.mergeAlbumsKey) as? Bool ?? true
		// Stored as the tag so it is a single value — a format and a bitrate
		// that could disagree would be two settings pretending to be one.
		audioQuality =
			defaults.string(forKey: Self.audioQualityKey).flatMap(AudioQuality.parse) ?? .default
		libraryMode = defaults.string(forKey: Self.libraryModeKey)
		albumSorts = defaults.dictionary(forKey: Self.albumSortsKey) as? [String: String] ?? [:]
		// `double(forKey:)` answers 0 for an absent key, which would be a cap
		// of nothing rather than the default — the same trap the merge switch
		// above avoids with `object(forKey:)`.
		offlineMode = defaults.bool(forKey: Self.offlineKey)
		let storedCap = defaults.object(forKey: Self.cacheCapKey) as? Double
		cacheCapBytes = storedCap.map { Int64($0) } ?? Self.defaultCacheCapBytes
	}
}
